### PR TITLE
Add katello-installer

### PR DESCRIPTION
**Pull request by ekohl**
This adds all the modules, hooks, checks and scenarios from katello-installer. It's still WIP because it's untested and the disk size check should be rewritten as a hook that's only applied to the katello scenarios. https://github.com/theforeman/foreman-installer/pull/242 is an early attempt at this.

* [ ] Reach consensus about proxy certificate generation
* [x] Merge https://github.com/Katello/katello-installer/pull/725
* [x] Merge https://github.com/Katello/katello-installer/pull/727
* [x] Rebase

### CODE REVIEW

File: Puppetfile [hunk 1]
```diff
@@ -1,15 +1,27 @@
 forge 'https://forgeapi.puppetlabs.com'
 
 # Dependencies
-mod 'puppetlabs/mysql',         '>= 4.0.0'
-mod 'puppetlabs/postgresql',    '>= 4.8.0'
+mod 'katello/candlepin',             :git => 'https://github.com/theforeman/puppet-candlepin'
```
Comment (ehelms) [inline Puppetfile]: Would it be cleaner to further split these inside this file with comments to make obvious katello specific module dependencies?

Comment (ekohl) [reply Puppetfile]: I considered this and had it in an earlier version. On the other hand we have one bunch of modules and don't split those out so not sure what the benefit would be.

**Comment by ekohl** [comment]
@mmoll for Debian packaging I'm considering just dropping the katello scenarios and hooks but leaving the modules in.

**Comment by ehelms** [comment]
Could this get a rebase and updated code?

**Comment by ehelms** [comment]
I'm thinking we:

 * Announce that we are going to freeze katello-installer in one week to PRs
 * Update this PR with any updated code at that point
 * Test and review for 2-3 days
 * Merge it!

The packaging proposals from our discussions sound like they are:

 * Includes all modules in both RPM and Deb
 * Drop Katello scenarios on Deb

As for RPM packaging options, here are some thoughts:

 * Option 1: Update foreman-installer spec to provide foreman-installer-katello sub-package to replace the current foreman-installer-katello package and put this into Katello repos via comps
 * Option 2: No sub-package, Katello now simply requires foreman-installer and is more dependent upon foreman-installer releases. This may be rockier at the start, but would provide the cleanest release.
 * Option 3: Generate two source tarballs from foreman-installer repository, one for foreman and one for Katello to package separately

I realize we should probably send these to discourse for any further input. I am including them here first to figure out the road you want to take @ekohl

**Comment by ekohl** [comment]
I've been looking at various files in the katello-installer repository. There's
# Modules

The current `Puppetfile` will be added to foreman-installer's `Puppetfile`. That means they'll be part of the tarball. Here we need to decide if we want to filter them out in the Debian installer, but I'm leaning to not doing so to keep differences minimal.

# Configs
These are essentially 3 scenarios:
* foreman-proxy-certs (special case)
* foreman-proxy-content
* katello

These would be part of the tarball but in packaging we split them off. For RPM I'm leaning to a subpackage (@ehelms' option 1). My reasoning is that we'll want a subpackage for the scripts anyway. More on that later. For Debs we'd simply exclude them since they don't work due to missing packages.

# Hooks

Part of the tarballs. It is unknown how they'll behave in the Foreman context so they'll be stored separate from the foreman-installer hooks. For packaging I'd keep this in the katello subpackage and allow Debs to exclude them.

# scripts

* capsule-installer
Deprecation wrapper we can drop
* capsule-remove
Deprecation wrapper we can drop
* create-migration
Development helper script, doesn't need to be packaged. Needs to be extended because Foreman uses a slightly different format (`%Y` vs `%y`).
* foreman-proxy-certs-generate
Complex script that needs some changes due to its use of paths. Belongs in the katello subpackage and excluded on Debs.
* katello-certs-check
Can be moved into the repo. Belongs in the katello subpackage and excluded on Debs
* katello-installer
Deprecation wrapper we can drop

# Boilerplate
Most of the Foreman should already be there.

The most important part is rubocop. Converting the foreman-installer to use rubocop is a large refactor. I'm proposing to make that a phase 2 that can also happen after 1.21.

# Dependencies

katello-installer-base has a few requirements. These should all be moved into foreman-installer-katello.

**Comment by ehelms** [comment]
ACK to Modules
ACK to Configs
ACK to Hooks
ACK to Scripts -- lets just remove those from the kinstaller now

Rubocop -- that shouldn't be that bad to retrofit, I'm gonna run it out of curiosity

**Comment by ekohl** [comment]
Rubocop: I've done it and it's not hard, but it's a big PR that creates a lot of noise. Reviews will be easier if we split it up.

**Comment by ehelms** [comment]
If we borrow the same setup from katello-installer as our starting baseline I see the following after auto-correct. To me this seems reasonable to go ahead and add in a separate PR

8 files changed, 36 insertions(+), 28 deletions(-)

**Comment by ekohl** [comment]
So I forgot the checks:

* disk_size.rb
Potentially problematic because `/var/lib/mongodb` space isn't really relevant in the Foreman scenario. Probably needs to be converted into a hook (https://github.com/theforeman/foreman-installer/pull/242).

* dns.rb
Already present in Foreman

* env_variables.rb
Could make sense in the Foreman scenario as well

* hostname.rb
Already present in Foreman

* lang.rb
Makes sense in the Foreman scenario. We won't work with the C locale.

**Comment by ekohl** [comment]
Updated, but untested. Going to update packaging.

Do have a look at foreman-proxy-certs as it takes a very different approach than in katello-installer. IMHO it's much less hacky and you can actually run the command locally which can make development much smoother.

File: bin/foreman-proxy-certs-generate [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env ruby
```
Comment (ehelms) [inline bin/foreman-proxy-certs-generate:1]: I thin this is a risky change to be making here. I think the intent is in a good place but this adds risk, testing burden, and potential for compounding issues. I would prefer split this off to be made after the merge as a re-factor that is independent.

**Comment by ehelms** [comment]
How about naming the `katello` directory `katello-hooks` to make it clearer whats in there?

File: bin/foreman-proxy-certs-generate (hunk 2 above)
Comment (ekohl) [reply bin/foreman-proxy-certs-generate:1]: On the other hand, the way it's dealing with directories would already require significant changes. That made me consider this a lower risk.

**Comment by ekohl** [comment]
> How about naming the `katello` directory `katello-hooks` to make it clearer whats in there?

Then you'd end up with`katello-hooks/hooks`. It's named `katello` because the configs also consider that the `installer_dir`. Kafo always loads `$installer_dir/hooks`, even if it's not named in `hooks_dirs`.

File: bin/foreman-proxy-certs-generate (hunk 2 above)
Comment (ehelms) [reply bin/foreman-proxy-certs-generate:1]: Can you explain what you mean by "dealing with directories" and why that would require significant change givne this all works as is today?

**Comment by ehelms** [comment]
Sorry, still confused. You said Kafo will auto load `$installer_dir/hooks` but if it's named `katello-hooks/hooks` then the installer won't autoload it? That sounds like a good thing and gives us more control while being more explicit over the naming.

**Comment by ekohl** [comment]
`katello-hooks/hooks` will work, but it's somewhat redundant IMHO. No strong opinion since my follow up project will be making sure the foreman hooks are compatible and can be used in katello. At that point we can move to `/usr/share/foreman-installer/katello-hooks`, removing at least one layer.

**Comment by ehelms** [comment]
Seems `--disable-system-checks` is missing:

```
fatal: [centos7-katello-nightly]: FAILED! => changed=true 
  cmd: foreman-installer -v --scenario katello  --disable-system-checks --foreman-admin-password changeme
  delta: '0:00:04.281919'
  end: '2018-12-12 01:07:00.278244'
  msg: non-zero return code
  rc: 1
  start: '2018-12-12 01:06:55.996325'
  stderr: |-
    ERROR: Unrecognised option '--disable-system-checks'
  
    See: 'foreman-installer --help'
  stderr_lines:
  - 'ERROR: Unrecognised option ''--disable-system-checks'''
  - ''
  - 'See: ''foreman-installer --help'''
```

**Comment by ehelms** [comment]
Another one:

```
/usr/share/foreman-installer/checks/hostname.rb:53:in `<main>': undefined local variable or method `fqdn' for main:Object (NameError)
Your system does not meet configuration criteria
```

**Comment by ekohl** [comment]
> Seems `--disable-system-checks` is missing:

I suspect this has something to do with very early stage hooks (boot): they haven't read the scenario yet so when you haven't set up the installer scenario, it's an issue. I'll think about this. Perhaps it'd be a good candidate to merge with the build in `--skip-checks-i-know-better`

**Comment by ehelms** [comment]
Testing a proxy I hit (and kinda what I was hinting at with "risky" changes to generate):

```
fatal: [pipeline-proxy-nightly-centos7 -> 192.168.121.81]: FAILED! => changed=true 
  cmd:
  - foreman-proxy-certs-generate
  - --foreman-proxy-fqdn
  - pipeline-proxy-nightly-centos7.war.example.com
  - --certs-tar
  - /root/pipeline-proxy-nightly-centos7.war.example.com.tar.gz
  delta: '0:00:21.311210'
  end: '2018-12-12 14:06:51.023463'
  msg: non-zero return code
  rc: 6
  start: '2018-12-12 14:06:29.712253'
  stderr: |-
    Debug: Loading facts from /usr/share/foreman-insta: 0 00:00:00
    Debug: Facter: 219                                : 0 0.0/s 00:00:01
    Debug: Automatically imported certs::config from c: 0 0.0/s 00:00:02
    Debug: Prefetching yum resources for package      : 0/18, 0%, 0.0/s, elapsed: 00:00:02
    Package[katello-certs-tools]                      : 0/18, 0%, 0.0/s, elapsed: 00:00:02
    File[/etc/pki/katello]                            : 1/18, 5%, 0.0/s, elapsed: 00:00:02
    File[/etc/pki/katello/certs]                      : 2/18, 11%, 0.0/s, elapsed: 00:00:02
    File[/etc/pki/katello/private]                    : 3/18, 16%, 0.0/s, elapsed: 00:00:02
    File[/etc/pki/katello/private/katello-default-ca.p: 4/18, 22%, 0.0/s, elapsed: 00:00:02
    Ca[katello-default-ca]                            : 5/18, 27%, 0.0/s, elapsed: 00:00:02
    Ca[katello-server-ca]                             : 6/18, 33%, 1.4/s, elapsed: 00:00:03, ETA: 00:00:08
    [31m Could not set 'present' on ensure: No such file or directory @ rb_sysopen - /etc/pki/katello-certs-tools/certs/katello-default-ca.crt (file: /usr/share/foreman-installer/modules/certs/manifests/ca.pp, line: 56)
    [0m[31m/opt/puppetlabs/puppet/lib/ruby/2.4.0/fileutils.rb:1289:in `initialize'
```

**Comment by ekohl** [comment]
> Testing a proxy I hit (and kinda what I was hinting at with "risky" changes to generate):

Does this work on nightly? Could also be a puppet-certs regression we're hitting.

**Comment by ehelms** [comment]
Appears to be fine on current nightly:

```
TASK [foreman_proxy_content : Generate Certs] **********************************
changed: [pipeline-proxy-nightly-centos7 -> 192.168.121.14] => changed=true 
  cmd:
  - foreman-proxy-certs-generate
  - --foreman-proxy-fqdn
  - pipeline-proxy-nightly-centos7.n31.example.com
  - --certs-tar
  - /root/pipeline-proxy-nightly-centos7.n31.example.com.tar.gz
  delta: '0:00:39.441146'
  end: '2018-12-12 13:54:58.044111'
  rc: 0
  start: '2018-12-12 13:54:18.602965'
  stderr: |-
    tput: unknown terminal "unknown"
    tput: unknown terminal "unknown"
    Debug: Loading facts from /usr/share/foreman-insta: 0 00:00:00
    Info: Facter: loading custom facts from /usr/share: 0 0.0/s 00:00:01
    }.                                                : 0 0.0/s 00:00:02
        Global Data Provider (hiera configuration vers: 0 0.0/s 00:00:04
    Debug: Prefetching yum resources for package      : 0/18, 0%, 0.0/s, elapsed: 00:00:04
    Package[katello-certs-tools]                      : 0/18, 0%, 0.0/s, elapsed: 00:00:05
    File[/etc/pki/katello]                            : 1/18, 5%, 0.0/s, elapsed: 00:00:05
    File[/etc/pki/katello/certs]                      : 2/18, 11%, 0.0/s, elapsed: 00:00:05
    File[/etc/pki/katello/private]                    : 3/18, 16%, 0.0/s, elapsed: 00:00:05
    File[/etc/pki/katello/private/katello-default-ca.p: 4/18, 22%, 0.0/s, elapsed: 00:00:05
    Ca[katello-default-ca]                            : 5/18, 27%, 0.0/s, elapsed: 00:00:05
    Ca[katello-server-ca]                             : 6/18, 33%, 0.0/s, elapsed: 00:00:05
    File[/root/ssl-build/KATELLO-TRUSTED-SSL-CERT]    : 7/18, 38%, 0.0/s, elapsed: 00:00:05
    Cert[pipeline-proxy-nightly-centos7.n31.example.co: 8/18, 44%, 0.0/s, elapsed: 00:00:05
    Cert[pipeline-proxy-nightly-centos7.n31.example.co: 9/18, 50%, 0.7/s, elapsed: 00:00:07, ETA: 00:00:12
    Cert[pipeline-proxy-nightly-centos7.n31.example.co: 10/18, 55%, 0.7/s, elapsed: 00:00:09, ETA: 00:00:12
    Cert[pipeline-proxy-nightly-centos7.n31.example.co: 11/18, 61%, 0.7/s, elapsed: 00:00:11, ETA: 00:00:09
    Cert[pipeline-proxy-nightly-centos7.n31.example.co: 12/18, 66%, 0.8/s, elapsed: 00:00:14, ETA: 00:00:07
    Cert[pipeline-proxy-nightly-centos7.n31.example.co: 13/18, 72%, 0.9/s, elapsed: 00:00:16, ETA: 00:00:05
    Cert[pipeline-proxy-nightly-centos7.n31.example.co: 14/18, 77%, 1.0/s, elapsed: 00:00:18, ETA: 00:00:04
    Cert[pipeline-proxy-nightly-centos7.n31.example.co: 15/18, 83%, 0.4/s, elapsed: 00:00:21, ETA: 00:00:06
    Cert[pipeline-proxy-nightly-centos7.n31.example.co: 16/18, 88%, 0.4/s, elapsed: 00:00:23, ETA: 00:00:04
    Exec[generate /root/pipeline-proxy-nightly-centos7: 17/18, 94%, 0.5/s, elapsed: 00:00:25, ETA: 00:00:02
    Done                                              : 18/18, 100%, 0.5/s, elapsed: 00:00:26
    Done                                              : 18/18, 100%, 0.5/s, elapsed: 00:00:26
```

**Comment by ekohl** [comment]
I've added TODOs at the top so we don't forget. Any PR that's merged in katello-installer should be added there until we cherry pick it here as well.

**Comment by ekohl** [comment]
Catalog compilation fails on non-RH. Submitted https://github.com/theforeman/puppet-katello/pull/272 to improve that.

**Comment by ehelms** [comment]
There is a slight upgrade issue with moving the RPM packaging to `noreplace` on the configuration file. That being that we need to add special migrations to handle the changes needed in the config file itself. Here is the current config file after an upgrade:

```
:name: Katello
:description: Install Foreman with Katello
:enabled: true
:log_dir: /var/log/foreman-installer
:store_dir: ''
:log_name: katello.log
:log_level: DEBUG
:no_prefix: false
:mapping: {}
:answer_file: /etc/foreman-installer/scenarios.d/katello-answers.yaml
:installer_dir: /usr/share/katello-installer-base
:module_dirs:
- /usr/share/foreman-installer/modules
- /usr/share/katello-installer-base/modules
:colors: true
:color_of_background: :dark
:hook_dirs:
- /usr/share/katello-installer-base/hooks
:custom: {}
:low_priority_modules: []
:verbose_log_level: info
:skip_puppet_version_check: false
:default_values_dir: /etc/foreman-installer/scenarios.d
:parser_cache_path:
- /usr/share/foreman-installer/parser_cache/foreman.yaml
- /usr/share/katello-installer-base/parser_cache/katello.yaml
:hiera_config: /usr/share/foreman-installer/config/foreman-hiera.conf
:order:
- certs
- foreman
- katello
- foreman_proxy
- foreman_proxy::plugin::pulp
- foreman_proxy_content
- puppet
```

Note the kept references to `katello-installer-base` directory.

**Comment by ehelms** [comment]
I believe this change is missing from whats here currently (https://github.com/Katello/katello-installer/pull/727) as I hit upgrade errors.

**Comment by ehelms** [comment]
Cheers! Completed upgrade test scenario successfully (still needs https://github.com/theforeman/foreman-installer/pull/268#issuecomment-447046051)

**Comment by ehelms** [comment]
Tests are green here now. I think you can drop WIP? Is there a final to-do list before merge you can to tackle?

**Comment by ekohl** [comment]
See the top

File: bin/foreman-proxy-certs-generate (hunk 2 above)
Comment (ekohl) [reply bin/foreman-proxy-certs-generate:1]: Everything in foreman-installer is able to run locally and in a portable way. This allows us to run it in CI as well as production. The changes to allow the current script are about as big as my new modification where my new version is a lot more forward compatible by being less hacky.

**Review by ehelms — APPROVED** [review]
Passed all install and upgrade tests for me! Bravo!

**Pull request merged**